### ISSUE TITLE
pre_push command fails when there's deleted refs on a remote with huge amount of refs

### ISSUE THREAD

**Issue by zxm5010** [github/git-lfs#1355]
Hey,

I think this bug is related to issue: https://github.com/github/git-lfs/issues/1014

Repro:
1. Have thousands remote branches on a repo
2. Delete any remote branch
3. Getting error 

> Error scanning for Git LFS files

Problem: func revListArgsRefVsRemote() in scanner.go will append tons of refs which introduce the "argument list too long" error and fails the push

```
func revListArgsRefVsRemote(refTo, remoteName string) []string {
	// We need to check that the locally cached versions of remote refs are still
	// present on the remote before we use them as a 'from' point. If the
	// server implements garbage collection and a remote branch had been deleted
	// since we last did 'git fetch --prune', then the objects in that branch may
	// have also been deleted on the server if unreferenced.
	// If some refs are missing on the remote, use a more explicit diff

	cachedRemoteRefs, _ := git.CachedRemoteRefs(remoteName)
	actualRemoteRefs, _ := git.RemoteRefs(remoteName)

	// Only check for missing refs on remote; if the ref is different it has moved
	// forward probably, and if not and the ref has changed to a non-descendant
	// (force push) then that will cause a re-evaluation in a subsequent command anyway
	missingRefs := NewStringSet()
	for _, cachedRef := range cachedRemoteRefs {
		found := false
		for _, realRemoteRef := range actualRemoteRefs {
			if cachedRef.Type == realRemoteRef.Type && cachedRef.Name == realRemoteRef.Name {
				found = true
				break
			}
		}
		if !found {
			missingRefs.Add(cachedRef.Name)
		}
	}

	if len(missingRefs) > 0 {
		// Use only the non-missing refs as 'from' points
		ret := []string{refTo, "--not"}
		for _, cachedRef := range cachedRemoteRefs {
			if !missingRefs.Contains(cachedRef.Name) {
				ret = append(ret, fmt.Sprintf("refs/remotes/%v/%v", remoteName, cachedRef.Name))
			}
		}
		return ret
	} else {
		// Safe to use cached
		return []string{refTo, "--not", "--remotes=" + remoteName}
	}

}

```

Hack-fix to temp solve this one is, I have to do git fetch --prune before any push when there's a deleted ref on remote.

**Comment by technoweenie** [comment]
Thanks for the report!

**Comment by MaciejKucia** [comment]
This is a duplicate of #1318

I personally use a custom build where only following code is used:
```
// Safe to use cached
        return []string{refTo, "--not", "--remotes=" + remoteName}
```

**Comment by technoweenie** [comment]
@MaciejKucia Ah, thanks, I knew it was already reported. Add a `bug` label to that one too.

**Comment by sinbad** [comment]
It's tricky to resolve this without also re-introducing the bug the explicit remote diff fixes. A potential coping mechanism would be to detect the case of an argument overrun and either automatically run `git fetch --prune`, or fail and prompt the user to do it. I'd probably err on the side of the latter because of the principle of least surprise.

**Comment by technoweenie** [comment]
Fixed in #1359, released in v1.3.0 :metal:
